Fault tolerance of Google calendar integration

Occasional internal server error with this stacktrace:
```
Dec  2 18:12:44 raspberrypizero run.sh[445]: 2019-12-02 18:12:44,801 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-worker-thread-5) HTTP Request to /index.html failed, error id: 0140602e-3098-45cd-aa4b-9f601395a5c4-1: org.jboss.resteasy.spi.UnhandledException: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:496)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:363)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:109)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatchRequestContext(VertxRequestHandler.java:84)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.lambda$handle$0(VertxRequestHandler.java:71)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.lang.Thread.run(Thread.java:748)
Dec  2 18:12:44 raspberrypizero run.sh[445]: Caused by: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.steffeleffe.HtmlResource_Subclass.hello(HtmlResource_Subclass.zig:94)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at sun.reflect.GeneratedMethodAccessor15.invoke(Unknown Source)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.lang.reflect.Method.invoke(Method.java:498)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:151)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.MethodInjectorImpl.lambda$invoke$3(MethodInjectorImpl.java:122)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:110)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:122)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:594)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:468)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:421)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:363)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:423)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:391)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invoke$1(ResourceMethodInvoker.java:365)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:365)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:477)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011... 14 more
Dec  2 18:12:44 raspberrypizero run.sh[445]: Caused by: java.net.SocketException: Network is unreachable (connect failed)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.net.PlainSocketImpl.socketConnect(Native Method)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
Dec  2 18:12:44 raspberrypizero run.sh[445]: #011at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at java.net.Socket.connect(Socket.java:589)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:666)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.NetworkClient.doConnect(NetworkClient.java:175)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.http.HttpClient.openServer(HttpClient.java:463)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.http.HttpClient.openServer(HttpClient.java:558)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:264)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:367)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:162)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:104)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at org.steffeleffe.calendarimport.GoogleCalendarImporter.getEvents(GoogleCalendarImporter.kt:63)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at org.steffeleffe.calendarimport.GoogleCalendarImporter.importCalender(GoogleCalendarImporter.kt:170)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at org.steffeleffe.calendarservice.CalendarService.getAllCalendars(CalendarService.kt:19)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at org.steffeleffe.HtmlResource.hello(HtmlResource.kt:29)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at org.steffeleffe.HtmlResource_Subclass.hello$$superaccessor1(HtmlResource_Subclass.zig:114)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at org.steffeleffe.HtmlResource_Subclass$$function$$1.apply(HtmlResource_Subclass$$function$$1.zig:29)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at io.smallrye.metrics.interceptors.TimedInterceptor.timedCallable(TimedInterceptor.java:95)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at io.smallrye.metrics.interceptors.TimedInterceptor.timedMethod(TimedInterceptor.java:70)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at io.smallrye.metrics.interceptors.TimedInterceptor_Bean.intercept(TimedInterceptor_Bean.zig:236)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011at org.steffeleffe.HtmlResource_Subclass.hello(HtmlResource_Subclass.zig:72)
Dec  2 18:12:45 raspberrypizero run.sh[445]: #011... 36 more
```

--- a/src/main/kotlin/org/steffeleffe/CalendarResource.kt
+++ b/src/main/kotlin/org/steffeleffe/CalendarResource.kt
@@ -17,7 +17,7 @@ class CalendarResource {
     @GET
     @Produces(MediaType.APPLICATION_JSON)
     fun getAllCalendars(): List<CalendarEvent> {
-        return calendarService.getAllCalendars(5)
+        return calendarService.getAllCalendars()
     }
 
 }

--- a/src/main/kotlin/org/steffeleffe/CssResource.kt
+++ b/src/main/kotlin/org/steffeleffe/CssResource.kt
@@ -31,7 +31,7 @@ open class CssResource {
     @Timed(name = "timed", description = "A measure of how long it takes to fetch css page.", unit = MetricUnits.MILLISECONDS)
     @Produces(MediaType.TEXT_PLAIN)
     open fun hello(): String {
-        val allCalendars = calendarService.getAllCalendars(5)
+        val allCalendars = calendarService.getAllCalendars()
         val usedTimes = getAllTimesUsedInEvents(allCalendars)
 
         val cssBuilder = CSSBuilder().apply {

--- a/src/main/kotlin/org/steffeleffe/HtmlResource.kt
+++ b/src/main/kotlin/org/steffeleffe/HtmlResource.kt
@@ -26,7 +26,7 @@ open class HtmlResource {
     @Timed(name = "timed", description = "A measure of how long it takes to fetch index page.", unit = MetricUnits.MILLISECONDS)
     @Produces(MediaType.TEXT_HTML)
     open fun hello(): String {
-        val allCalendars = calendarService.getAllCalendars(5)
+        val allCalendars = calendarService.getAllCalendars()
         val createHTML = createHTML(true, true)
         createHTML.head {
             styleLink("/style.css")

--- a/src/main/kotlin/org/steffeleffe/calendarservice/CalendarService.kt
+++ b/src/main/kotlin/org/steffeleffe/calendarservice/CalendarService.kt
@@ -1,24 +1,21 @@
 package org.steffeleffe.calendarservice
 
-import org.steffeleffe.calendarimport.GoogleCalendarImporter
 import javax.enterprise.context.ApplicationScoped
-import javax.enterprise.inject.Default
-import javax.inject.Inject
 
 @ApplicationScoped
-open class CalendarService {
+open class CalendarService() {
 
-    private val googleCalendarImporter = GoogleCalendarImporter()
+    private val cache = CalenderEventCache().cache
 
-    fun getAllCalendars(numberOfDays: Int): List<CalendarEvent> {
+    fun getAllCalendars(): List<CalendarEvent> {
 
         val joinedList = mutableListOf<CalendarEvent>()
 
-        joinedList.addAll(googleCalendarImporter.importCalender("primary", numberOfDays))
-        joinedList.addAll(googleCalendarImporter.importCalender("3eq4uqnkhcgipgkdrrhs7ec6e4@group.calendar.google.com", numberOfDays))
-        joinedList.addAll(googleCalendarImporter.importCalender("rikke.vangsted@gmail.com", numberOfDays))
-        joinedList.addAll(googleCalendarImporter.importCalender("66aglhcacpcpupnhh9fian0a1g@group.calendar.google.com", numberOfDays))
-        joinedList.addAll(googleCalendarImporter.importCalender("hdn3t11kjru1fs823pee8g9bso@group.calendar.google.com", numberOfDays))
+        joinedList.addAll(cache.get("primary"))
+        joinedList.addAll(cache.get("3eq4uqnkhcgipgkdrrhs7ec6e4@group.calendar.google.com"))
+        joinedList.addAll(cache.get("rikke.vangsted@gmail.com"))
+        joinedList.addAll(cache.get("66aglhcacpcpupnhh9fian0a1g@group.calendar.google.com"))
+        joinedList.addAll(cache.get("hdn3t11kjru1fs823pee8g9bso@group.calendar.google.com"))
 
         return joinedList
     }

--- a/src/main/kotlin/org/steffeleffe/calendarservice/CalenderEventCache.kt
+++ b/src/main/kotlin/org/steffeleffe/calendarservice/CalenderEventCache.kt
@@ -1,0 +1,39 @@
+package org.steffeleffe.calendarservice
+
+import com.google.common.cache.CacheBuilder
+import com.google.common.cache.CacheLoader
+import com.google.common.cache.LoadingCache
+import com.google.common.util.concurrent.ListenableFutureTask
+import org.steffeleffe.calendarimport.GoogleCalendarImporter
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+
+class CalenderEventCache {
+
+    private val numberOfDays = 5
+
+    private val googleCalendarImporter = GoogleCalendarImporter()
+
+    inner class CalendarEventCacheLoader : CacheLoader<String, List<CalendarEvent>>() {
+        override fun load(key: String): List<CalendarEvent> { // no checked exception
+            return getEvents(key)
+        }
+
+        override fun reload(key: String, oldValue: List<CalendarEvent>): ListenableFutureTask<List<CalendarEvent>> {
+            val task: ListenableFutureTask<List<CalendarEvent>> = ListenableFutureTask.create { getEvents(key) }
+            Executors.newSingleThreadExecutor().execute(task)
+            return task
+        }
+
+        private fun getEvents(key: String) : List<CalendarEvent> = googleCalendarImporter.importCalender(key, numberOfDays)
+
+    }
+
+    val cache: LoadingCache<String, List<CalendarEvent>> = CacheBuilder.newBuilder()
+            .maximumSize(1000)
+            .refreshAfterWrite(1, TimeUnit.MINUTES)
+            .build(CalendarEventCacheLoader())
+
+
+}

